rllib/core/rl_trainer/trainer_runner.py

```diff
@@ -100,19 +106,103 @@ def __init__(
     def is_local(self) -> bool:
         return self._is_local
 
-    def update(self, batch: MultiAgentBatch) -> List[Mapping[str, Any]]:
-        """Do a gradient based update to the RLTrainer(s) maintained by this TrainerRunner.
+    def fit(
+        self,
+        batch: MultiAgentBatch,
+        *,
+        minibatch_size: int,
+        num_iters: int,
+        reduce_fn: Optional[Callable[[ResultDict], ResultDict]] = _reduce_mean_results,
+    ) -> Mapping[str, Any]:
+        """Do `num_iters` minibatch updates given the original batch.
+
+        Given a batch of episodes you can use this method to take more
+        than one backward pass on the batch. The same minibatch_size and num_iters gets
+        will be used for all module ids (previously known as policies) in the
+        multiagent batch
 
         Args:
             batch: The data to use for the update.
+            minibatch_size: The size of the minibatch to use for each update.
+            num_iters: The number of complete passes over all the sub-batches
+                in the input multi-agent batch.
+            reduce_fn: See `update()` documenation for more details.
+
+        Returns:
+            A dictionary of results summarizing the statistics of the updates.
+        """
+
+        start = {mid: 0 for mid in batch.policy_batches.keys()}
+        num_covered_epochs = {mid: 0 for mid in batch.policy_batches.keys()}
+        results = []
+        # TODO (Kourosh): One data transfer is probably better than many for each mini
+        # batch. How should we do this?
+        # loop until the number of passes through all modules batches reaches the
+        # num_iters
+        while min(num_covered_epochs.values()) < num_iters:
+            minibatch = {}
+            for module_id, module_batch in batch.policy_batches.items():
+                s = start[module_id]  # start
+                e = s + minibatch_size  # end
+
+                samples_to_concat = []
+                # cycle through the batch until we have enough samples
+                while e >= len(module_batch):
```

Comment:
you duplicate module_batches multiple times to make a mini_batch large enough?
this block of code is actually very confusing ...

Reply:
yes, generally speaking, if number of samples within each policy batch is un-even and skewed, you need to make sure that they are sharded almost equally when you pass them to the RLTrainer. Say one policy1 have 100 samples, and policy2 has 20 samples. but when I want to pass a minibatch size of 40, from policy1 I will select 0-39, but from policy2 I select 0-19+0-19 to make up 40 samples. This is the problem of sharding across policies, I couldn't figure out an easy-to-understand sharding strategy. But I want to think about this more when I think about the double batch communication overhead, they are very relevant.

Reply:
I see. maybe make a util function out of this. so then it's pluggable. and we can write and compare a few different schemes.

Reply:
btw, not sure if you want to add these as comments in the code. that would help a lot.

Reply:
will update taking into account the suggestions 👍

Reply:
ok I am creating a utility MiniBatchIterator that can be re-used once we move this iteration to sharded batches inside RLTrainer as well. Thanks for the suggestion. I really like the design break down.

Reply:
👍